2019-12-naming-part1/naming.md

```diff
@@ -0,0 +1,234 @@
+# Naming: Climbing Towards Abstraction
+
+Your code and its artifacts are strewn with names. You name repositories, files, packages, classes, functions, variables, and on and on. The names you choose determine the quality of the conversations you have with your programmer peers.
+
+On the flip side of that bargain, your understanding of libraries and frameworks, i.e. other people's code, starts with names. Names form the substrate of APIs, delineating their boundaries, clarifying their inputs and outputs, and defining how they're used.
+
+If you can trust that a library works, you need look no further than its API docs to understand it. Ideally, API docs are the only point of contact you have with a third-party library. And they are, essentially, a bunch of names.
+
+Just as a writer grinds through word choices, a programmer must struggle with names. As the joke goes, the two hardest things in computer science are naming, cache invalidation, and off-by-one errors. No matter the programming language, no matter whether you're operating from a functional or object-oriented perspective, no matter if the language is statically or dynamically typed, names are a ubiquitous and critical part of your code. Cache invalidation may not matter in your application, but names sure do. The quality of your work depends on the quality of your names.
+
+When writing code, you spend a lot of time burrowed in the minutiae of specifics. You ply your craft at the mercy of unbending syntax rules &mdash; rules that will bring a million-line codebase to its knees over a single missing character. You write small functions that, given an exact input, produce an exact output. Every time.
+
+But naming demands that you wrest your mind from granular, concrete details and think more generally, more categorically.
+
+The linguist Sam Hayakawa, in his book [Language in Thought and Action](https://www.amazon.com/Language-Thought-Action-S-I-Hayakawa/dp/0156482401/), describes a model of language he calls the Abstraction Ladder. On this conceptual ladder, the bottommost rung holds the most concrete idea. The topmost rung is occupied by a decidedly abstract, intangible concept. In between, ideas get more abstract the higher up they are. The Abstraction Ladder is straightforward to understand, but studying it leads to broad and profound insights.
+
+<img src="images/abstraction-ladder.svg" width="100%" alt="The Abstraction Ladder" />
+
+Consider the diagram above. Molly is on the second rung. She is far from a static entity &mdash; she's a being in process, constantly changing, composed of numerous subsystems that are represented on the bottom rung. From a cognitive, linguistic perspective, however, Molly seems like a single entity. Indeed, associating a name with a living being is often the first utterance of a toddler learning to speak - "Mama."
```

Comment:
This is very minor but I'd put an em dash here "toddler learning to speak -- "Mama."

Reply:
Resolved in 953c87eedef5e23f55300d93cb56f52f381f0f6c.